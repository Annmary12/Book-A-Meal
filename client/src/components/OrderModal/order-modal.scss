@import '../../../assets/scss/_variables';

.header {
  text-align: center;
  @media only screen and (max-width: 600px) {}

  @media only screen and (min-width: 600px) {}

  @media only screen and (min-width: 768px) {}

  @media only screen and (min-width: 992px) {}

  @media only screen and (min-width: 1200px) {
    font-size: 24px;
  }
}

.overlay {
  position: fixed;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  background-color: #0c0c0cbf;
}

.body{
  position: absolute;
  top: 144px;
  left: 340px;
  right: 340px;
  background: white;
  overflow: auto;
  border-radius: 4px;
  outline: none;
  padding: 20px;

  @media only screen and (max-width: 600px) {
    font-size: 14px;
  }

  @media only screen and (min-width: 600px) {}

  @media only screen and (min-width: 768px) {
    left: 108px;
    right: 108px;
  }

  @media only screen and (min-width: 992px) {
    left: 177px;
    right: 177px;
  }

  @media only screen and (min-width: 1200px) {
    left: 266px;
    right: 266px;
  }

  @media only screen and (max-width: 600px) {
    top: 32px;
    left: 20px;
    right: 20px;
  }
}

.content{
  display: flex;

  @media only screen and (max-width: 600px) {
    flex-direction: column;
  }
}

.imageContainer{
  display: flex;
  flex-basis: 40%;
  img {
    width: 232px;
    max-height: 231px;
    margin: auto;

    @media only screen and (max-width: 600px) {
      width: 64%;
      height: 75%;
    }

    @media only screen and (min-width: 600px) {}

    @media only screen and (min-width: 768px) {
      width: 172px;
    }

    @media only screen and (min-width: 992px) {
      width: 190px;
    }

    @media only screen and (min-width: 1200px) {

    }

    @media only screen and (max-width: 600px) {

    }
  }

  @media only screen and (max-width: 600px) {
    margin-bottom: 10px;
  }

}

.detailArea {
  flex-basis: 60%;
}

.mealImage{
  border-radius: 5px;
  width: 194px;
  @media only screen and (max-width: 600px) {
    margin: auto;
    width: 157px;
  }
}

.input {
  box-sizing: border-box;
  width: 100%;
  border: 2px solid #c1b8b8;
  border-style: solid;
  border-radius: 5px;
  font-size: 15px;
  margin: auto;
  margin-bottom: -4px;
  padding: 10px;
  height: 33px;

  &:focus{
    outline: coral auto 4px
  }

  &:disabled{
    border-color:transparent;
    padding: 0px;
  }
}

.deliverAddressInput {
  @extend .input;
  height: 68px;
}

.quantityBtn {
  border-radius: 5px;
  border: 1px solid #cac5c5f5;
  padding: 1px 7px 1px 7px;
  text-align: center;
  background-color: #eaeaeafa;
  cursor: pointer;
}

.quantityCtrl {
  display: flex;
  justify-content: space-between;
  width: 88px;
  margin-left: 8px;
}

.quantityArea{
  display: flex;
}

.info {
  margin-bottom: 10px;
}

.btn {
  padding: 7px 5px;
  font-size: 17px;
  border-radius: 5px;
  min-width: 81px;
  cursor: pointer;
  outline: none;
}

.successBtn {
  @extend .btn;
  color: white;
  background-color: $green-btn-color;
  border: 2px solid darken($green-btn-color, 5%);
  &:hover{
    background-color: darken($green-btn-color, 5%);
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  }

  &:disabled, &[disabled]{
    background-color: darken($green-btn-color, 5%);
    &:hover{
      box-shadow: none;
      cursor: none;
    }
  }
}

.closeModalBtn {
  @extend .btn;
  border: 2px solid #3f3e3e;
  background-color: transparent;
  color: #3f3e3e;
}
.btns {
  display: flex;
  justify-content: space-between;
  margin-top: 10px;
  @media only screen and (max-width: 600px) {
    display: flex;
    justify-content: space-between;
  }
}
.fadeInLeft {
  :global {
    animation: fadeInLeft .4s ease-in;
  }
}
.cancelBtn {
  @extend .btn;
  border: 2px solid #b94421;
  background-color: transparent;
  color: #b94421;
}


.errorInput{
  @extend .input;
  border-color: #de1e1e;
  :global{
    animation: shake .4s ease-in .2s;
  }
}

.errorDeliveryAddressInput{
  @extend .deliverAddressInput;
  border-color: #de1e1e;
  :global{
    animation: shake .4s ease-in .2s;
  }
}

.errorMessage {
  color: #de1e1e;
  font-size: 13px;
  margin-top: -10px;
  margin-bottom: 10px;
}

@media only screen and (max-width: 600px) {}

@media only screen and (min-width: 600px) {}

@media only screen and (min-width: 768px) {}

@media only screen and (min-width: 992px) {}

@media only screen and (min-width: 1200px) {}