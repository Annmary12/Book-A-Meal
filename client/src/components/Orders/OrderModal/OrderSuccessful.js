import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import checkMark from '../../../../assets/img/check-mark.svg';
import orderStyles from './order-modal.scss';

const OrderSuccessful = ({ handleClose }) => (
  <div style={{ textAlign: 'center' }}>
    <h3>Order placed successfully</h3>
    <img src={checkMark} alt="check-mark" width="50" />
    <p>
      <Link to="/orders">View order history</Link>
    </p>
    <p>
      Please note that order update and cancellation
      can only be done 15 minutes after the order is placed.
    </p>

    <div style={{ paddingTop: '35px' }}>
      <button
        className={orderStyles.closeModalBtn}
        onClick={handleClose}
      >
      close
      </button>
    </div>
  </div>
);

OrderSuccessful.propTypes = {
  handleClose: PropTypes.func.isRequired
};

export default OrderSuccessful;
