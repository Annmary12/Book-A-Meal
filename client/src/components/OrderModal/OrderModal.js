import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import * as Yup from 'yup';
import classnames from 'classnames';
import OrderSuccessful from './OrderSuccessful.js';
import OrderService from '../../services/api/orders.js';
import localizeNum from '../../utils/localizeNum.js';
import requestErrorHandler from '../../utils/requestErrorHandler.js';
import orderStyles from './order-modal.scss';
import loadingCube from '../../../assets/img/loading-cube.svg';

const initialState = {
  quantity: 1,
  fields: {
    deliveryAddress: '',
    phoneNumber: ''
  },
  errors: {},
  touched: {},
  detailsEntered: false,
  orderRequestIsProcessing: false,
  orderSuccessful: false
};

const schema = Yup.object().shape({
  deliveryAddress: Yup.string()
    .min(10, 'Delivery address should be at least 10 characters')
    .max(100, 'Delivery address should be less than 100 characters')
    .matches(
      /^[a-z0-9 (),.'-]+$/i,
      "Address can only contain letters, numbers and any of these ( , . ' - )"
    )
    .required('Delivery address is required'),
  phoneNumber: Yup.string()
    .matches(
      /^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,9}$/i,
      'Phone number is invalid.'
    )
    .min(11)
    .required('Phone number is required')
});

/**
 * @class OrderModal
 */
class OrderModal extends Component {
  static propTypes = {
    meal: PropTypes.shape({
      id: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired
    }).isRequired,
    isOpen: PropTypes.bool.isRequired,
    handleClose: PropTypes.func.isRequired
  }

  state = initialState

  placeOrder = () => {
    this.setState({ orderRequestIsProcessing: true });
    const orderDetails = {
      mealId: this.props.meal.id,
      quantity: this.state.quantity,
      phoneNumber: this.state.fields.phoneNumber,
      deliveryAddress: this.state.fields.deliveryAddress
    };

    OrderService.placeOrder(orderDetails)
      .then(() => {
        this.setState({
          orderRequestIsProcessing: false,
          orderSuccessful: true
        });
      })
      .catch((error) => {
        this.setState({
          orderRequestIsProcessing: false,
          orderSuccessful: false
        });
        requestErrorHandler(error);
      });
  }

  handleChange = (event) => {
    const targetName = event.target.name;

    this.setState({
      fields: {
        ...this.state.fields,
        [event.target.name]: event.target.value
      }
    }, () => this.validate(targetName));
  }

  handleBlur = (event) => {
    this.validate(event.target.name);
  }

  validate = (targetName) => {
    schema.validate(this.state.fields, { abortEarly: false }).then(() => {
      const { errors } = this.state;
      delete errors[targetName];

      this.setState({
        errors
      });
    }).catch((error) => {
      const errors = {};
      error.inner.forEach((err) => {
        errors[err.path] = err.message;
      });
      this.setState(prevState => ({
        touched: {
          ...prevState.touched,
          [targetName]: true
        },
        errors
      }));
    });
  }

  handleClose = () => {
    this.setState(initialState);
    this.props.handleClose();
  }

  proceedToConfirm = () => {
    schema.validate(this.state.fields, { abortEarly: false }).then(() => {
      this.setState({ detailsEntered: true });
    }).catch((error) => {
      const errors = {};
      error.inner.forEach((err) => {
        errors[err.path] = err.message;
      });
      const touched = Object.keys(this.state.fields)
        .reduce((acc, field) => ({ ...acc, [field]: true }), {});
      this.setState({
        touched,
        errors
      });
    });
  }

  increaseQuantity = () => {
    this.setState(prevState => ({ quantity: prevState.quantity + 1 }));
  }

  reduceQuantity = () => {
    this.setState((prevState) => {
      if (prevState.quantity > 1) {
        return {
          quantity: prevState.quantity - 1
        };
      }
    });
  }
  goBackToEdit = () => {
    this.setState({ detailsEntered: false });
  }
  /**
   * @return {JSX} React JSX
   */
  render() {
    const { meal, isOpen } = this.props;
    const { handleClose } = this;
    const btnsClass = classnames({
      [orderStyles.btns]: true,
      [orderStyles.flipInLeft]: this.state.detailsEntered
    });

    return (
      <ReactModal
        isOpen={isOpen}
        onRequestClose={handleClose}
        overlayClassName="react-modal-overlay"
        className="react-modal-body"
        closeTimeoutMS={150}
      >
        {
      this.state.orderSuccessful ?
        <OrderSuccessful
          handleClose={handleClose}
        /> :
        <div>
          <h3 className={orderStyles.header}>
            {
              this.state.detailsEntered ?
              'Confirm order' :
              'Order information'
            }
          </h3>
          <div className={orderStyles.content}>
            <div className={orderStyles.imageContainer}>
              <img className={orderStyles.mealImage} src={meal.img} alt="meal" />
            </div>
            <div className={orderStyles.detailArea}>
              <div className={orderStyles.info}>
                <b>Title: </b>{meal.title}
              </div>
              <div className={orderStyles.info}>
                <b>Price: </b>&#8358;{localizeNum(meal.price * this.state.quantity)}
              </div>
              <div
                className={classnames(orderStyles.info, orderStyles.quantityArea)}
              >
                <b>Quantity: </b>
                <span className={orderStyles.quantityCtrl}>
                  {
                    !this.state.detailsEntered &&
                    <button
                      className={orderStyles.quantityBtn}
                      onClick={this.reduceQuantity}
                    >
                      -
                    </button>
                  }
                  {this.state.quantity}
                  {
                    !this.state.detailsEntered &&
                    <button
                      className={orderStyles.quantityBtn}
                      onClick={this.increaseQuantity}
                    >
                      +
                    </button>
                  }
                </span>
              </div>
              <div className={orderStyles.info}>
                <b>Phone number: </b><br />
                <input
                  className={
                    this.state.errors.phoneNumber && this.state.touched.phoneNumber ?
                    orderStyles.errorInput :
                    orderStyles.input
                  }
                  value={this.state.fields.phoneNumber}
                  onChange={this.handleChange}
                  onBlur={this.handleBlur}
                  disabled={this.state.detailsEntered ? 'disabled' : ''}
                  name="phoneNumber"
                />
                {
                  this.state.errors.phoneNumber && this.state.touched.phoneNumber &&
                  <span
                    className={orderStyles.errorMessage}
                  >
                    {this.state.errors.phoneNumber}
                  </span>
                }
              </div>
              <div className={orderStyles.info}>
                <b>Delivery address: </b><br />
                <textarea
                  value={this.state.fields.deliveryAddress}
                  onChange={this.handleChange}
                  name="deliveryAddress"
                  onBlur={this.handleBlur}
                  disabled={this.state.detailsEntered ? 'disabled' : ''}
                  maxLength="101"
                  className={
                    this.state.errors.deliveryAddress && this.state.touched.deliveryAddress ?
                    orderStyles.errorDeliveryAddressInput :
                    orderStyles.deliverAddressInput

                  }
                />
                {
                  this.state.errors.deliveryAddress && this.state.touched.deliveryAddress &&
                  <span
                    className={orderStyles.errorMessage}
                  >
                    {this.state.errors.deliveryAddress}
                  </span>
                }
              </div>
            </div>
          </div>
          <div>
            {
              this.state.detailsEntered ?
                <div className={btnsClass}>
                  {!this.state.orderRequestIsProcessing &&

                    <button
                      className={orderStyles.cancelBtn}
                      onClick={this.goBackToEdit}
                    >
                    Edit details
                    </button>
                  }

                  <button
                    className={orderStyles.successBtn}
                    onClick={this.placeOrder}
                    disabled={this.state.orderRequestIsProcessing ? 'disabled' : ''}
                  >
                    {
                      !this.state.orderRequestIsProcessing ? 'Place order' :
                      <span>
                        Processing
                        <img style={{ paddingLeft: '5px' }} src={loadingCube} width="20" alt="loading" />
                      </span>
                    }
                  </button>
                </div> :
                <div className={btnsClass}>
                  <button
                    className={orderStyles.cancelBtn}
                    onClick={handleClose}
                  >
                    Cancel
                  </button>
                  <button
                    className={orderStyles.successBtn}
                    onClick={this.proceedToConfirm}
                  >
                    Continue
                  </button>
                </div>
              }
          </div>
        </div>
      }
      </ReactModal>
    );
  }
}

export default OrderModal;
