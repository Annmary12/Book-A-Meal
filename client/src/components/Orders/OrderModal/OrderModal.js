import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import OrderSuccessful from './OrderSuccessful.js';
import OrderService from '../../../services/api/orders.js';
import requestErrorHandler from '../../../utils/requestErrorHandler.js';
import OrderInfoForm from '../OrderInfoForm.js';

/**
 * @class OrderModal
 *
 * @param {Object} orderDetails
 */
class OrderModal extends Component {
  static propTypes = {
    meal: PropTypes.shape({
      id: PropTypes.string.isRequired,
      img: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired
    }).isRequired,
    isOpen: PropTypes.bool.isRequired,
    handleClose: PropTypes.func.isRequired
  }

  state = {
    orderRequestIsProcessing: false,
    orderSuccessful: false
  }

  placeOrder = (orderDetails) => {
    this.setState({ orderRequestIsProcessing: true });

    return OrderService.placeOrder(orderDetails)
      .then(() => {
        this.setState({
          orderRequestIsProcessing: false,
          orderSuccessful: true
        });
      })
      .catch((error) => {
        this.setState({
          orderRequestIsProcessing: false,
          orderSuccessful: false
        });
        requestErrorHandler(error);
      });
  }

  /**
   * @return {JSX} React JSX
   */
  render() {
    const { meal, isOpen, handleClose } = this.props;
    const { orderSuccessful, orderRequestIsProcessing } = this.state;

    return (
      <ReactModal
        isOpen={isOpen}
        onRequestClose={handleClose}
        overlayClassName="react-modal-overlay"
        className="react-modal-body"
        closeTimeoutMS={150}
      >
        {
          orderSuccessful ?
            <OrderSuccessful
              handleClose={handleClose}
            /> :
            <OrderInfoForm
              meal={meal}
              handleSubmit={this.placeOrder}
              proceedText="Confirm order"
              headerText="Confirm order information"
              isProcessing={orderRequestIsProcessing}
              handleClose={handleClose}
            />
        }
      </ReactModal>
    );
  }
}

export default OrderModal;
