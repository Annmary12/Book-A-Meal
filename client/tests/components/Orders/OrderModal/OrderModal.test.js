import React from 'react';
import OrderModal from '../../../../src/components/Orders/OrderModal/OrderModal';

const props = {
  meal: {
    id: 'adf92d',
    img: 'picture.png',
    title: 'rice and stew',
    description: 'white rice and stew',
    price: 3289
  },
  isOpen: false,
  handleClose: jest.fn()
}
describe('OrderModal component', () => {
  const wrapper = shallow(
    <OrderModal
      {...props}
    />
  );
  it('should match snapshot if orderSuccessful is false', () => {
    wrapper.instance().setState({ orderSuccessful: false });

    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot if orderSuccessful is true', () => {
    wrapper.instance().setState({ orderSuccessful: true });

    expect(wrapper).toMatchSnapshot();
  });

  describe('placeOrder method', () => {
    wrapper.instance().placeOrder({});

    it('should set orderRequestIsProcessing to false', () => {
      const { orderRequestIsProcessing } = wrapper.instance().state;

      expect(orderRequestIsProcessing).toBeFalsy();
    })
    it('should set orderSuccessful to true', () => {
      const { orderSuccessful } = wrapper.instance().state;

      expect(orderSuccessful).toBeTruthy();
    })
  })
});