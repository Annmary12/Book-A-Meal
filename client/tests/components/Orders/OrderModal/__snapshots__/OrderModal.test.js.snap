// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderModal component should match snapshot if orderSuccessful is false 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="react-modal-body"
  closeTimeoutMS={150}
  isOpen={false}
  onRequestClose={[MockFunction]}
  overlayClassName="react-modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <OrderInfoForm
    deliveryAddress=""
    handleClose={[MockFunction]}
    handleSubmit={[Function]}
    headerText="Confirm order information"
    isProcessing={false}
    meal={
      Object {
        "description": "white rice and stew",
        "id": "adf92d",
        "img": "picture.png",
        "price": 3289,
        "title": "rice and stew",
      }
    }
    phoneNumber=""
    proceedText="Confirm order"
    quantity={1}
  />
</Modal>
`;

exports[`OrderModal component should match snapshot if orderSuccessful is true 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="react-modal-body"
  closeTimeoutMS={150}
  isOpen={false}
  onRequestClose={[MockFunction]}
  overlayClassName="react-modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <OrderSuccessful
    handleClose={[MockFunction]}
  />
</Modal>
`;
